Parameter Pack Error: extra arguments at positions #2, #3, #4, #5, #6, #7, #8, #9, #10, #11, #12, #13, #14, #15, #16, #17, #18, #19, #20, #21, #22, #23, #24, #25, #26, #27, #28, #29, #30, #31, #32, #33, #34, #35, #36, #37, #38, #39, #40, #41, #42, #43, #44, #45, #46, #47, #48, #49, #50...#65536

<!--
This repository tracks issues related to the implementation of the Swift
compiler, standard library, runtime, and tools that provide IDE support for
Swift (e.g. code completion). If the bug relates to the implementation of a
proprietary (closed-source) Apple framework such as UIKit, SwiftUI, Combine,
etc., please report it to https://feedbackassistant.apple.com instead.
-->

**Description**
<!-- Describe clearly and concisely what the bug is. -->
Compiling the following code produces a ridiculous error message complaining about extra arguments at tens of thousands of different positions.

Error given is (real error contains full 65536 positions)
```
test.swift:12:30: error: extra arguments at positions #2, #3, #4, #5, #6, #7, #8, #9, #10, #11, #12, #13, #14, #15, #16, #17, #18, #19,
#20, #21, #22, #23, #24, #25, #26, #27, #28, #29, #30, #31, #32, #33, #34, #35, #36, #37, #38, #39, #40, #41, #42, #43, #44,
#45, #46, #47, #48, #49, #50, #51, #52, #53, #54, #55, #56, #57, #58, #59, #60, #61, #62, #63, #64, #65, #66, #67, #68, #69, 
#70, #71, #72, #73, #74, #75, #76, #77, #78, #79, #80, #81, #82, #83, #84, #85, #86, #87, #88, #89, #90, #91, #92, #93, #94,
#95, #96, #97, #98, #99, #100, #101, #102, #103, #104, #105, #106, #107, #108, #109, #110, #111, #112, #113, #114, #115, #116, 
#117, #118, #119, #120, #121, #122, #123, #124, #125, #126, #127, #128, #129, #130, #131, #132, #133, #134, #135, #136, #137, 
#138, #139, #140, #141, #142, #143, #144, #145, #146, #147, #148, #149, #150, #151, #152, #153, #154, #155, #156, #157, #158, 
#159, #160, #161, #162, #163, #164, #165, #166, #167, #168, #169, #170, #171, #172, #173, #174, #175, #176, #177, #178, #179, 
#180, #181, #182, #183, #184, #185, #186, #187, #188, #189, #190, #191, #192, #193, #194, #195, #196, #197, #198, #199, #200, 
#201...#65534, #65535, #65536 in call
    previousValue = function(repeat each trigger)
                             ^
test.swift:7:7: note: 'function' declared here
  let function: Function
      ^
```
**Steps to reproduce**
<!--
Explain how to reproduce the problem (in steps if seen fit) and include either
an inline test case (preferred) or a project that reproduces it. Consider
reducing the sample to the smallest amount of code possible — a smaller test
case is easier to reason about and more appealing to contributors.
-->

``` swift
struct Cached<each Trigger: Equatable, Value>{
  typealias Function = (repeat each Trigger) -> Value

  var previousTriggers: (repeat each Trigger)?
  var previousValue: Value?

  let function: Function

  mutating func go(_ trigger: repeat each Trigger) -> Value {
    let triggers = (repeat each trigger)

    previousValue = function(repeat each trigger)
    previousTriggers = triggers

    return previousValue!
  }
}

var cached = Cached {a, b in
  let a: Int = a
  let b: Int = b

  return a + b
}

print(cached.go(2, 3))

```

**Expected behaviour**
<!-- Describe what you expected to happen. -->
 I would expect the compiler to infer the number of arguments from the function provided, or give an appropriate error message.

**Environment**
- Swift compiler version info: swift-driver version: 1.87.1 Apple Swift version 5.9 (swiftlang-5.9.0.128.2 clang-1500.0.40.1) Target: x86_64-apple-macosx14.0
- Xcode version info Xcode 15.0 Build version 15A5219j
- Deployment target: x86_64-apple-macosx14.0

<!-- Add any other context about the problem as appropriate. -->
<!-- **Additional context** -->

--- a/lib/AST/ParameterPack.cpp
+++ b/lib/AST/ParameterPack.cpp
@@ -453,8 +453,11 @@ PackType::getExpandedGenericArgs(ArrayRef<GenericTypeParamType *> params,
 }
 
 PackType *PackType::getSingletonPackExpansion(Type param) {
-  assert(param->isParameterPack() || param->is<PackArchetypeType>());
-  return get(param->getASTContext(), {PackExpansionType::get(param, param)});
+  SmallVector<Type, 2> rootParameterPacks;
+  param->getTypeParameterPacks(rootParameterPacks);
+  assert(rootParameterPacks.size() >= 1);
+  auto count = rootParameterPacks[0];
+  return get(param->getASTContext(), {PackExpansionType::get(param, count)});
 }
 
 CanPackType CanPackType::getSingletonPackExpansion(CanType param) {

--- a/lib/Sema/CSDiagnostics.h
+++ b/lib/Sema/CSDiagnostics.h
@@ -449,7 +449,9 @@ class SameShapeExpansionFailure final : public FailureDiagnostic {
 public:
   SameShapeExpansionFailure(const Solution &solution, Type lhs, Type rhs,
                             ConstraintLocator *locator)
-      : FailureDiagnostic(solution, locator), lhs(lhs), rhs(rhs) {}
+      : FailureDiagnostic(solution, locator),
+        lhs(resolveType(lhs)),
+        rhs(resolveType(rhs)) {}
 
   bool diagnoseAsError() override;
 };

--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -13624,32 +13624,33 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifySameShapeConstraint(
         auto argLoc =
             loc->castLastElementTo<LocatorPathElt::ApplyArgToParam>();
 
-        if (type1->getAs<PackArchetypeType>() &&
-            type2->getAs<PackArchetypeType>())
+        if (type1->is<PackArchetypeType>() &&
+            type2->is<PackArchetypeType>())
           return recordShapeMismatchFix();
 
-        auto argPack = type1->getAs<PackType>();
-        auto paramPack = type2->getAs<PackType>();
-
-        if (!(argPack && paramPack))
-          return SolutionKind::Error;
+        auto numArgs = (shape1->is<PackType>()
+                        ? shape1->castTo<PackType>()->getNumElements()
+                        : 1);
+        auto numParams = (shape2->is<PackType>()
+                          ? shape2->castTo<PackType>()->getNumElements()
+                          : 1);
 
         // Tailed diagnostic to explode tuples.
         // FIXME: This is very similar to
         // 'cannot_convert_single_tuple_into_multiple_arguments'; can we emit
         // both of these in the same place?
-        if (argPack->getNumElements() == 1) {
-          if (argPack->getElementType(0)->is<TupleType>() &&
-              paramPack->getNumElements() >= 1) {
+        if (numArgs == 1) {
+          if (type1->is<TupleType>() &&
+              numParams >= 1) {
             return recordShapeFix(
                 DestructureTupleToMatchPackExpansionParameter::create(
-                    *this, paramPack, loc),
-                /*impact=*/2 * paramPack->getNumElements());
+                    *this,
+                    (type2->is<PackType>()
+                     ? type2->castTo<PackType>()
+                     : PackType::getSingletonPackExpansion(type2)), loc),
+                /*impact=*/2 * numParams);
           }
         }
-
-        auto numArgs = shape1->castTo<PackType>()->getNumElements();
-        auto numParams = shape2->castTo<PackType>()->getNumElements();
 
         // Drops `ApplyArgToParam` and left with `ApplyArgument`.
         path.pop_back();

--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -1807,7 +1807,14 @@ static ConstraintSystem::TypeMatchResult matchCallArguments(
         }
 
         auto *argPack = PackType::get(cs.getASTContext(), argTypes);
-        auto *argPackExpansion = PackExpansionType::get(argPack, argPack);
+        auto argPackExpansion = [&]() {
+          if (argPack->getNumElements() == 1 &&
+              argPack->getElementType(0)->is<PackExpansionType>()) {
+            return argPack->getElementType(0)->castTo<PackExpansionType>();
+          }
+
+          return PackExpansionType::get(argPack, argPack);
+        }();
 
         auto firstArgIdx =
             argTypes.empty() ? paramIdx : parameterBindings[paramIdx].front();

--- a/lib/Sema/ConstraintSystem.cpp
+++ b/lib/Sema/ConstraintSystem.cpp
@@ -855,25 +855,7 @@ Type ConstraintSystem::openUnboundGenericType(GenericTypeDecl *decl,
     result = DC->mapTypeIntoContext(result);
   }
 
-  return result.transform([&](Type type) -> Type {
-    // Although generic parameters are declared with just `each`
-    // their interface types introduce a pack expansion which
-    // means that the solver has to extact generic argument type
-    // variable from Pack{repeat ...} and drop that structure to
-    // make sure that generic argument gets inferred to a pack type.
-    if (auto *packTy = type->getAs<PackType>()) {
-      assert(packTy->getNumElements() == 1);
-      auto *expansion = packTy->getElementType(0)->castTo<PackExpansionType>();
-      auto *typeVar = expansion->getPatternType()->castTo<TypeVariableType>();
-      assert(typeVar->getImpl().getGenericParameter() &&
-             typeVar->getImpl().canBindToPack());
-      return typeVar;
-    }
-
-    if (auto *expansion = dyn_cast<PackExpansionType>(type.getPointer()))
-      return openPackExpansionType(expansion, replacements, locator);
-    return type;
-  });
+  return result;
 }
 
 static void checkNestedTypeConstraints(ConstraintSystem &cs, Type type,
@@ -1019,18 +1001,6 @@ Type ConstraintSystem::openType(Type type, OpenedTypeMap &replacements,
                     locator)},
                 tuple->getASTContext());
           }
-        }
-      }
-
-      // While opening variadic generic types that appear in other types
-      // we need to extract generic parameter from Pack{repeat ...} structure
-      // that gets introduced by the interface type, see
-      // \c openUnboundGenericType for more details.
-      if (auto *packTy = type->getAs<PackType>()) {
-        if (auto expansionTy = packTy->unwrapSingletonPackExpansion()) {
-          auto patternTy = expansionTy->getPatternType();
-          if (patternTy->isTypeParameter())
-            return openType(patternTy, replacements, locator);
         }
       }
 
@@ -3926,6 +3896,14 @@ struct TypeSimplifier {
       // Transform the count type, ignoring any active pack expansions.
       auto countType = expansion->getCountType().transform(
           TypeSimplifier(CS, GetFixedTypeFn));
+
+      if (!countType->is<PackType>() &&
+          !countType->is<PackArchetypeType>()) {
+        SmallVector<Type, 2> rootParameterPacks;
+        countType->getTypeParameterPacks(rootParameterPacks);
+        if (!rootParameterPacks.empty())
+          countType = rootParameterPacks[0];
+      }
 
       // If both pattern and count are resolves, let's just return
       // the pattern type for `transformWithPosition` to take care

--- a/test/Constraints/issue-67906.swift
+++ b/test/Constraints/issue-67906.swift
@@ -1,0 +1,25 @@
+// RUN: %target-typecheck-verify-swift -disable-availability-checking
+
+struct G<each T>: Sequence {
+  typealias Element = Int
+  typealias Iterator = [Int].Iterator
+
+  consuming func makeIterator() -> Iterator {
+    fatalError()
+  }
+}
+
+// expected-note@+1 {{in call to function 'foo'}}
+func foo<each T>(_: repeat each T) -> G<repeat each T> {
+  .init()
+}
+
+// expected-error@+2 {{for-in loop requires '(repeat each T) -> G<repeat each T>' to conform to 'Sequence'}}
+// expected-error@+1 {{generic parameter 'each T' could not be inferred}}
+for a in foo {
+  print(a)
+}
+
+for a in foo() {
+  print(a)
+}

--- a/test/Constraints/pack-expansion-expressions.swift
+++ b/test/Constraints/pack-expansion-expressions.swift
@@ -139,7 +139,8 @@ func tupleExpansion<each T, each U>(
   _ = zip(repeat each tuple1, with: repeat each tuple1.element) // legacy syntax
 
   _ = zip(repeat each tuple1, with: repeat each tuple2)
-  // expected-error@-1 {{global function 'zip(_:with:)' requires the type packs 'repeat each T' and 'repeat each U' have the same shape}}
+  // expected-error@-1 {{global function 'zip(_:with:)' requires the type packs 'each T' and 'each U' have the same shape}}
+  // expected-error@-2 {{pack expansion requires that 'each U' and 'each T' have the same shape}}
 
   _ = forward(repeat each tuple3)
 }
@@ -349,37 +350,36 @@ func test_pack_expansions_with_closures() {
 func test_pack_expansion_specialization(tuple: (Int, String, Float)) {
   struct Data<each T> {
     init(_: repeat each T) {} // expected-note 4 {{'init(_:)' declared here}}
-    init(vals: repeat each T) {} // expected-note {{'init(vals:)' declared here}}
-    init<each U>(x: Int, _: repeat each T, y: repeat each U) {} // expected-note 3 {{'init(x:_:y:)' declared here}}
+    init(vals: repeat each T) {}
+    init<each U>(x: Int, _: repeat each T, y: repeat each U) {}
   }
 
   _ = Data<Int>() // expected-error {{missing argument for parameter #1 in call}}
   _ = Data<Int>(0) // Ok
   _ = Data<Int, String>(42, "") // Ok
-  _ = Data<Int>(42, "") // expected-error {{extra argument in call}}
+  _ = Data<Int>(42, "") // expected-error {{pack expansion requires that 'Int' and 'Int, String' have the same shape}}
   _ = Data<Int, String>((42, ""))
-  // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}} {{25-26=}} {{32-33=}}
+  // expected-error@-1 {{initializer expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}} {{25-26=}} {{32-33=}}
   _ = Data<Int, String, Float>(vals: (42, "", 0))
-  // expected-error@-1 {{value pack expansion at parameter #0 expects 3 separate arguments; remove extra parentheses to change tuple into separate arguments}} {{38-39=}} {{48-49=}}
+  // expected-error@-1 {{pack expansion requires that 'Int, String, Float' and '(Int, String, Int)' have the same shape}}
   _ = Data<Int, String, Float>((vals: 42, "", 0))
-  // expected-error@-1 {{value pack expansion at parameter #0 expects 3 separate arguments; remove extra parentheses to change tuple into separate arguments}} {{32-33=}} {{48-49=}}
+  // expected-error@-1 {{initializer expects 3 separate arguments; remove extra parentheses to change tuple into separate arguments}} {{32-33=}} {{48-49=}}
   _ = Data<Int, String, Float>(tuple)
-  // expected-error@-1 {{value pack expansion at parameter #0 expects 3 separate arguments}}
+  // expected-error@-1 {{initializer expects 3 separate arguments}}
   _ = Data<Int, String, Float>(x: 42, tuple)
-  // expected-error@-1 {{value pack expansion at parameter #1 expects 3 separate arguments}}
+  // expected-error@-1 {{pack expansion requires that 'Int, String, Float' and '(Int, String, Float)' have the same shape}}
   _ = Data<Int, String, Float>(x: 42, tuple, y: 1, 2, 3)
-  // expected-error@-1 {{value pack expansion at parameter #1 expects 3 separate arguments}}
+  // expected-error@-1 {{pack expansion requires that 'Int, String, Float' and '(Int, String, Float)' have the same shape}}
   _ = Data<Int, String, Float>(x: 42, (42, "", 0), y: 1, 2, 3)
-  // expected-error@-1 {{value pack expansion at parameter #1 expects 3 separate arguments}} {{39-40=}} {{49-50=}}
+  // expected-error@-1 {{pack expansion requires that 'Int, String, Float' and '(Int, String, Int)' have the same shape}}
 
   struct Ambiguity<each T> {
     func test(_: repeat each T) -> Int { 42 }
-    // expected-note@-1 {{value pack expansion at parameter #0 expects 3 separate arguments}}
+    // expected-note@-1 {{'test' declared here}}
     func test(_: repeat each T) -> String { "" }
-    // expected-note@-1 {{value pack expansion at parameter #0 expects 3 separate arguments}}
   }
 
-  _ = Ambiguity<Int, String, Float>().test(tuple) // expected-error {{no exact matches in call to instance method 'test'}}
+  _ = Ambiguity<Int, String, Float>().test(tuple) // expected-error {{instance method 'test' expects 3 separate arguments}}
 }
 
 // rdar://107280056 - "Ambiguous without more context" with opaque return type + variadics
@@ -621,44 +621,40 @@ do {
 // https://github.com/apple/swift/issues/66393
 do {
   struct S<each T> {
-    var property: (repeat each T) -> Void { // expected-note 4 {{'property' declared here}}
+    var property: (repeat each T) -> Void {
       get {}
     }
 
-    func method(_: repeat each T) {} // expected-note 4 {{'method' declared here}}
+    func method(_: repeat each T) {}
   }
   S<Int, Bool>().method((5, true))
-  // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+  // expected-error@-1 {{pack expansion requires that 'Int, Bool' and '(Int, Bool)' have the same shape}}
 
   S<Int, Bool>().method((5, true, 6))
-  // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+  // expected-error@-1 {{pack expansion requires that 'Int, Bool' and '(Int, Bool, Int)' have the same shape}}
 
   S<Int, Bool>().property((5, true))
-  // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+  // expected-error@-1 {{cannot pass value pack expansion to non-pack parameter of type 'repeat each T'}}
 
   S<Int, Bool>().property((5, true, 6))
-  // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+  // expected-error@-1 {{cannot pass value pack expansion to non-pack parameter of type 'repeat each T'}}
 
   func foo<each U>(u: repeat each U) {
     S<repeat each U>().property((3, 4, 5))
-    // expected-error@-1 {{value pack expansion at parameter #0 expects 1 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+    // expected-error@-1 {{cannot pass value pack expansion to non-pack parameter of type 'repeat each T'}}
 
     // FIXME: The count of 'repeat each U' is not statically known, but error suggests that it is 1.
     S<repeat each U>().method((3, 4, 5))
-    // expected-error@-1 {{value pack expansion at parameter #0 expects 1 separate arguments; remove extra parentheses to change tuple into separate arguments}}
-    // FIXME: Bad diagnostics
-    // expected-error@-3 {{pack expansion requires that 'each U' and '_' have the same shape}}
-    // expected-error@-4 {{pack expansion requires that 'each U' and '_.RawValue' have the same shape}}
+    // expected-error@-1 {{pack expansion requires that 'each U' and '(Int, Int, Int)' have the same shape}}
 
     // FIXME: The count of '(Int, Int), repeat each U' is not statically known, but error suggests that it is 2.
     S<(Int, Int), repeat each U>().method((3, 4))
-    // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
-    // FIXME: Duplicate diagnostics
-    // expected-error@-3 2 {{pack expansion requires that 'each U' and '' have the same shape}}
+    // expected-error@-1 {{pack expansion requires that '(Int, Int), repeat each U' and '(Int, Int)' have the same shape}}
+    // expected-error@-2 {{pack expansion requires that '' and 'each U' have the same shape}}
 
     // FIXME: The count of '(Int, Int), repeat each U' is not statically known, but error suggests that it is 2.
     S<(Int, Int), repeat each U>().property((3, 4))
-    // expected-error@-1 {{value pack expansion at parameter #0 expects 2 separate arguments; remove extra parentheses to change tuple into separate arguments}}
+    // expected-error@-1 {{cannot pass value pack expansion to non-pack parameter of type 'repeat each T'}}
   }
 }
 

--- a/test/Constraints/variadic_generic_constraints.swift
+++ b/test/Constraints/variadic_generic_constraints.swift
@@ -73,5 +73,6 @@ func goodCallToZip<each T, each U>(t: repeat each T, u: repeat each U) where (re
 
 func badCallToZip<each T, each U>(t: repeat each T, u: repeat each U) {
   _ = zip(t: repeat each t, u: repeat each u)
-  // expected-error@-1 {{global function 'zip(t:u:)' requires the type packs 'repeat each T' and 'repeat each U' have the same shape}}
+  // expected-error@-1 {{global function 'zip(t:u:)' requires the type packs 'each T' and 'each U' have the same shape}}
+  // expected-error@-2 {{pack expansion requires that 'each U' and 'each T' have the same shape}}
 }

--- a/test/Constraints/variadic_generic_init.swift
+++ b/test/Constraints/variadic_generic_init.swift
@@ -1,0 +1,84 @@
+// RUN: %target-typecheck-verify-swift -disable-availability-checking
+
+// These test cases exercise variants of rdar://problem/112785081
+// and https://github.com/apple/swift/issues/68160.
+
+protocol P {}
+
+protocol Q {
+  associatedtype A: Q
+
+  var a: A { get }
+}
+
+struct S1<each T: Q>: P {
+  init(_: repeat each T) {}
+}
+
+func foo1a<each T: Q>(_ t: repeat each T) -> some P {
+  return S1(repeat each t)
+}
+
+func foo2a<each T: Q>(_ t: repeat each T) -> S1<repeat each T> {
+  return S1(repeat each t)
+}
+
+func foo3a<each T: Q>(_ t: repeat each T) -> some P {
+  return S1(repeat (each t).a)
+}
+
+func foo4a<each T: Q>(_ t: repeat each T) -> S1<repeat (each T).A> {
+  return S1(repeat (each t).a)
+}
+
+func foo1b<each T: Q>(_ t: repeat each T) -> some P {
+  return S1.init(repeat each t)
+}
+
+func foo2b<each T: Q>(_ t: repeat each T) -> S1<repeat each T> {
+  return S1.init(repeat each t)
+}
+
+func foo3b<each T: Q>(_ t: repeat each T) -> some P {
+  return S1.init(repeat (each t).a)
+}
+
+func foo4b<each T: Q>(_ t: repeat each T) -> S1<repeat (each T).A> {
+  return S1.init(repeat (each t).a)
+}
+
+struct S2<each T: Q>: P {
+  init(arg: (repeat each T)) {}
+}
+
+func bar1a<each T: Q>(_ t: repeat each T) -> some P {
+  return S2(arg: (repeat each t))
+}
+
+func bar2a<each T: Q>(_ t: repeat each T) -> S2<repeat each T> {
+  return S2(arg: (repeat each t))
+}
+
+func bar3a<each T: Q>(_ t: repeat each T) -> some P {
+  return S2(arg: (repeat (each t).a))
+}
+
+func bar4a<each T: Q>(_ t: repeat each T) -> S2<repeat (each T).A> {
+  return S2(arg: (repeat (each t).a))
+}
+
+func bar1b<each T: Q>(_ t: repeat each T) -> some P {
+  return S2.init(arg: (repeat each t))
+}
+
+func bar2b<each T: Q>(_ t: repeat each T) -> S2<repeat each T> {
+  return S2.init(arg: (repeat each t))
+}
+
+func bar3b<each T: Q>(_ t: repeat each T) -> some P {
+  return S2.init(arg: (repeat (each t).a))
+}
+
+func bar4b<each T: Q>(_ t: repeat each T) -> S2<repeat (each T).A> {
+  return S2.init(arg: (repeat (each t).a))
+}

--- a/test/Constraints/variadic_generic_types.swift
+++ b/test/Constraints/variadic_generic_types.swift
@@ -56,6 +56,7 @@ do {
     return Test<String, Int, repeat each T>("a", 42, repeat each v) // Ok
   }
 }
+
 // rdar://107479662 - variadic tuple of Sendable elements does not conform to Sendable
 do {
   struct Test<each T> : Sendable {
@@ -63,5 +64,16 @@ do {
   }
 
   struct TestProperty<T> : Sendable {
+  }
+}
+
+// https://github.com/apple/swift/issues/68160
+do {
+  struct G<each T, U> {
+    let f: (repeat Optional<each T>) -> U
+
+    init(f: @escaping (repeat Optional<each T>) -> U) {
+      self.f = f
+    }
   }
 }

--- a/validation-test/compiler_crashers_2_fixed/rdar113463759.swift
+++ b/validation-test/compiler_crashers_2_fixed/rdar113463759.swift
@@ -1,0 +1,16 @@
+// RUN: %target-swift-frontend -emit-ir %s -disable-availability-checking
+
+protocol SomeObjectProtocol {}
+
+struct SomeObject: SomeObjectProtocol {}
+
+protocol ObjectProviding {
+    associatedtype Object: SomeObjectProtocol
+    var object: Object { get }
+}
+
+struct ConformingObject<each B>: ObjectProviding {
+    var object: some SomeObjectProtocol {
+        SomeObject()
+    }
+}

